lib/model/presence.dart

```diff
@@ -0,0 +1,182 @@
+      case AppLifecycleState.resumed:
+        // > […] the default running mode for a running application that has
+        // > input focus and is visible.
+        result = await updatePresence(connection,
+          pingOnly: pingOnly,
+          status: PresenceStatus.active,
+          newUserInput: true);
+      case AppLifecycleState.inactive:
+        // > At least one view of the application is visible, but none have
+        // > input focus. The application is otherwise running normally.
+        // For example, we expect this state when the user is selecting a file
+        // to upload.
+        result = await updatePresence(connection,
+          pingOnly: pingOnly,
+          status: PresenceStatus.active,
+          newUserInput: false);
+    }
+    if (!pingOnly) {
+      _map = result.presences!;
+      notifyListeners();
+    }
+  }
+
+  void _poll() async {
+    assert(!_disposed);
+    while (true) {
+      // We put the wait upfront because we already have data when [start] is
+      // called; it comes from /register.
+      await Future<void>.delayed(serverPresencePingInterval);
+      if (_disposed) return;
+
+      await _maybePingAndRecordResponse(
+        SchedulerBinding.instance.lifecycleState, pingOnly: false);
+      if (_disposed) return;
+    }
+  }
+
+  bool _disposed = false;
+
+  @override
+  void dispose() {
+    _appLifecycleListener?.dispose();
+    _disposed = true;
+    super.dispose();
+  }
+
+  /// The [PresenceStatus] for [userId], or null if the user is offline.
+  PresenceStatus? presenceStatusForUser(int userId, {required DateTime utcNow}) {
+    final now = utcNow.millisecondsSinceEpoch ~/ 1000;
+    final perUserPresence = _map[userId];
+    if (perUserPresence == null) return null;
+    final PerUserPresence(:activeTimestamp, :idleTimestamp) = perUserPresence;
+
+    if (now - activeTimestamp <= serverPresenceOfflineThresholdSeconds) {
+      return PresenceStatus.active;
+    } else if (now - idleTimestamp <= serverPresenceOfflineThresholdSeconds) {
+      // The API doc is kind of confusing, but this seems correct:
+      //   https://chat.zulip.org/#narrow/channel/378-api-design/topic/presence.3A.20.22potentially.20present.22.3F/near/2202431
+      // TODO clarify that API doc
+      return PresenceStatus.idle;
+    } else {
+      return null;
+    }
+  }
+
+  void handlePresenceEvent(PresenceEvent event) {
+    // TODO(#1618)
+  }
+
+  /// In debug mode, controls whether presence requests are made.
+  ///
+  /// Outside of debug mode, this is always true and the setter has no effect.
+  static bool get debugEnable {
+    bool result = true;
+    assert(() {
+      result = _debugEnable;
+      return true;
+    }());
+    return result;
+  }
+  static bool _debugEnable = true;
+  static set debugEnable(bool value) {
+    assert(() {
+      _debugEnable = value;
+      return true;
+    }());
+  }
+
+  @visibleForTesting
+  static void debugReset() {
+    _debugEnable = true;
```

Comment:
nit:
```suggestion
  static void debugReset() {
    debugEnable = true;
```

That way it goes through the setter above, so it's clear to the compiler that this field never gets written outside debug mode.